x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/CompositeAggregationDataExtractor.java

```diff
@@ -168,20 +168,28 @@ private InputStream processAggs(Aggregations aggs) throws IOException {
         ));
         aggregationToJsonProcessor.process(aggs);
         ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-        final boolean hasAfterKey = afterKey != null && (afterKey.get(context.compositeAggDateHistogramGroupSourceName) instanceof Long);
+        final Long afterKeyTimeBucket = afterKey != null ? (Long)afterKey.get(context.compositeAggDateHistogramGroupSourceName) : null ;
         boolean cancellable = aggregationToJsonProcessor.writeAllDocsCancellable(
             timestamp -> {
                 if (isCancelled) {
                     // If we have not processed a single composite agg page yet and we are cancelled
                     // We should not process anything
-                    if (hasAfterKey == false) {
+                    if (afterKeyTimeBucket == null) {
                         return true;
                     }
+                    // We want to stop processing once a timestamp enters the next time bucket.
+                    // This could occur in any page. One benefit we have is that even though the paging order is not sorted
+                    // by max timestamp, our iteration of the page results is. So, once we cross over to the next bucket within
+                    // a given page, we know the previous bucket has been exhausted.
                     if (nextBucketOnCancel == 0L) {
-                        // If we have been cancelled, record the bucket above our latest timestamp
-                        // This indicates when we have completed the current bucket of this timestamp and thus will move to the next
-                        // date_histogram bucket
-                        nextBucketOnCancel = Intervals.alignToCeil(timestamp, interval);
+                        if (timestamp.equals(afterKeyTimeBucket)) {
```

Comment:
if `afterKeyTimeBucket` is the floor, then could this not be simplified to always be `nextBucketOnCancel = afterKeyTimeBucket + interval` ?

Reply:
This ALMOST works. 

Our considerations are:
- 1. We iterate a given page in order of the `max` timestamp
- 2. A `max` timestamp for a bucket will always be within that bucket and interval `max -> [bucket, bucket+interval)`
- 3. A page could have multiple `date_histogram` buckets
- 4. Two pages could have out of order `max` timestamps as the paging order is predicated on bucket floor + terms
- 5. A cancel could occur at any time

If we take considerations 4 and 5, the timestamp at the moment of cancellation could be within a subsequent bucket. This may be 1 or more date_histogram buckets above the current page. 

Lets ponder:
 - our interval is `1000` 
 - we are on page `1000`
 - The current page passed over into the next bucket with a timestamp `2500`
 - Cancel is handled
 - If we simply set the `nextBucketOnCancel = afterKeyTimeBucket + interval`, the cancellation would be `2000`
 - We erroneously end before finishing the current `date_histogram` bucket (which is now `2000`)

Having monotonically iterating `max` timestamps within a given page is nice as it allows us to know when we have cross over into the next page and ONCE we have cross over into the next page, we know that all `max` timestamps will be AT LEAST as big as the floor of the next bucket. So, we can say with certainty once cross the floor of the next bucket, we are done.

Reply:
Got it. But I think we can get away with `Intervals.alignToFloor(timestamp + interval, interval)`

Reply:
100%, I think this works. Will update. Fewer if statements == fewer bugs (usually).